asserts/ifacedecls.go

```diff
@@ -0,0 +1,187 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+type attrMatcher interface {
+	match(interface{}) error
+}
+
+// compileAttrMatcher compiles an attrMatcher derived from constraints,
+// context can be "top" or "map", alternatives have no influence on it
+func compileAttrMatcher(context string, constraints interface{}) (attrMatcher, error) {
+	switch x := constraints.(type) {
+	case map[string]interface{}:
+		return compileMapAttrMatcher(x)
+	case []interface{}:
+		// TODO: disallow nesting alternatives directly inside alernatives?
+		return compileAltAttrMatcher(context, x)
+	case string:
+		if context == "top" {
+			return nil, fmt.Errorf("first level of non alternative constraints must be a set of key-value contraints")
+		}
+		return compileRegexpAttrMatcher(x)
+	default:
+		return nil, fmt.Errorf("constraint must be a key-value map, regexp or a list of alternative constraints: %v", x)
+	}
+}
+
+type mapAttrMatcher map[string]attrMatcher
+
+func compileMapAttrMatcher(m map[string]interface{}) (attrMatcher, error) {
+	matcher := make(mapAttrMatcher)
+	for k, constraint := range m {
+		matcher1, err := compileAttrMatcher("map", constraint)
+		if err != nil {
+			return nil, fmt.Errorf("constraint for %q: %v", k, err)
+		}
+		matcher[k] = matcher1
+	}
+	return matcher, nil
+}
+
+func matchEntry(k string, matcher1 attrMatcher, v interface{}) error {
+	if v == nil {
+		return fmt.Errorf("%q has constraints but is unset", k)
+	}
+	if err := matcher1.match(v); err != nil {
+		return fmt.Errorf("%q mismatch: %v", k, err)
+	}
+	return nil
+}
+
+func matchList(matcher attrMatcher, l []interface{}) error {
+	for i, elem := range l {
+		if err := matcher.match(elem); err != nil {
+			return fmt.Errorf("element %d: %v", i, err)
+		}
+	}
+	return nil
+}
+
+func (matcher mapAttrMatcher) match(v interface{}) error {
+	switch x := v.(type) {
+	case map[string]interface{}: // top level looks like this
+		for k, matcher1 := range matcher {
+			if err := matchEntry(k, matcher1, x[k]); err != nil {
+				return err
+			}
+		}
+	case map[interface{}]interface{}: // nested maps look like this
+		for k, matcher1 := range matcher {
+			if err := matchEntry(k, matcher1, x[k]); err != nil {
+				return err
+			}
+		}
+	case []interface{}:
+		return matchList(matcher, x)
+	default:
+		return fmt.Errorf("cannot match key-value constraints against: %v", v)
+	}
+	return nil
+}
+
+type regexpAttrMatcher struct {
+	*regexp.Regexp
+}
+
+func compileRegexpAttrMatcher(s string) (attrMatcher, error) {
+	rx, err := regexp.Compile("^" + s + "$")
+	if err != nil {
+		return nil, fmt.Errorf("cannot compile %q: %v", s, err)
+	}
+	return regexpAttrMatcher{rx}, nil
+}
+
+func (matcher regexpAttrMatcher) match(v interface{}) error {
+	var s string
+	switch x := v.(type) {
+	case string:
+		s = x
+	case bool:
+		s = strconv.FormatBool(x)
+	case int:
+		s = strconv.Itoa(x)
```

Comment:
float64?

You know what's a shame.. the yaml package knows how to unmarshal the value as a string. That'd be the best thing. Instead of parsing back and forth, get the actual representation in the file. I think I've used a trick to do that somewhere else.. maybe we can pull it off for this logic.